File: gol.rb
#!/usr/bin/env ruby

class Gol
    def initialize(w=30, prob=0.5)
        @w = w
        @prob = prob
        @map = []
        (0...@w).each do |i|
            (0...@w).each do |j|
                @map << (rand(100) < @prob*100 ? 1 : 0)
            end
        end
    end

    def [](x, y)
        @map[x*@w + y] == 1 ? 1 : 0
    end

    def print_
        (0...@w).each do |i|
            (0...@w).each do |j|
                print " X"[self[i, j]].chr + " "
            end
            puts
        end
    end

    def neighbors(x, y)
        dirs = [[-1, -1], [0, -1], [1, -1],
                [-1,  0],          [1,  0],
                [-1,  1], [0,  1], [1,  1]]

        cnt = 0
        dirs.each do |d|
            cnt += ((x+d[0] >= 0 and y+d[1] >= 0 and\
                     x+d[0] < @w and y+d[1] < @w and\
                     self[x+d[0], y+d[1]] == 1) ? 1 : 0)
        end
        cnt
    end

    def next_state
        ns = []
        (0...@w).each do |i|
            (0...@w).each do |j|
                nc = self.neighbors(i, j)
                ns << ((self[i, j] == 1 ? (nc == 2 or nc == 3) : (nc == 3)) ?
                       1 : 0)
            end
        end

        @map = ns
    end
end

game = Gol.new
game.print_
while 1
    game.next_state
    game.print_
    puts "="*60
    sleep 0.1
end

File: gol.py
#!/usr/bin/env python

import random
import time

class gol:
    def __init__(self):
        self._w = 30          # map width
        self._prob = 0.5      # probability
        self._map = []
        for i in range(self._w):
            for j in range(self._w):
                self._map.append(random.randint(0, 100) < self._prob*100)

    def get(self, x, y):
        return self._map[x*self._w + y]

    def print_(self):
        for i in range(self._w):
            for j in range(self._w):
                print ' X'[self.get(i, j)],
            print ""

    def neighbors(self, x, y):
        dirs = [(-1, -1), (0, -1), (1, -1),
                (-1,  0),          (1,  0),
                (-1,  1), (0,  1), (1,  1)]

        cnt = 0
        for d in dirs:
            cnt += (x+d[0] >= 0 and y+d[1] >= 0 and\
                    x+d[0] < self._w and y+d[1] < self._w and\
                    self.get(x+d[0], y+d[1]))

        return cnt

    def next_state(self):
        ns = []
        for i in range(self._w):
            for j in range(self._w):
                nc = self.neighbors(i, j)
                ns.append((nc in [2, 3]) if self.get(i, j) else (nc == 3))

        self._map = ns


game = gol()
game.print_()

try:
    while 1:
        game.next_state()
        game.print_()
        print "="*60
        time.sleep(0.1)

except KeyboardInterrupt:
    print "bye"

